File: LeetCode/Problems_JAVA/02.AddTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode result = null;
        ListNode q = null;
        int carry = 0;
        while(l1!=null||l2!=null){
            int a,b;
            if(l1==null){
                a = 0;
            }else{
                a = l1.val;
                l1 = l1.next;
            }
            if(l2==null){
                b = 0;
            }else{
                b = l2.val;
                l2 = l2.next;
            }


            int sum = a + b + carry;
            carry = sum/10;
            sum = sum%10;
            if(result==null){
                result = new ListNode(sum);
                q = result;
            }else{
                q.next = new ListNode(sum);
                q = q.next;
            }
        }
        if(carry!=0){
            q.next = new ListNode(carry);
        }
        return result;
    }
}

File: LeetCode/Problems2024/02.AddTwoNumbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* p = l1;
        ListNode* q = l2;
        int carry = 0;
        ListNode* result = NULL;
        ListNode* r = NULL;
        while (p || q) {
            int a = 0, b = 0;
            if (p) {
                a = p->val;
                p = p->next;
            }
            if (q) {
                b = q->val;
                q = q->next;
            }
            int sum = b + a + carry;
            carry = sum / 10;
            sum = sum % 10;
            if (result == NULL) {
                result = new ListNode(sum);
                r = result;
            } else {
                r->next = new ListNode(sum);
                r = r->next;
            }
        }
        if (carry != 0) {
            r->next = new ListNode(carry);
        }
        return result;
    }
};
